File: C#/Binary Tree - DFS/1372. Longest ZigZag Path in a Binary Tree.cs
/* 1372. Longest ZigZag Path in a Binary Tree
Difficulty: Medium
Submission Link: https://leetcode.com/problems/longest-zigzag-path-in-a-binary-tree/submissions/1388943045/ */

public class Solution {
    public int LongestZigZag(TreeNode root) {
        return Math.Max(DFS(root.left, true, 0), DFS(root.right, false, 0));
    }

    int DFS(TreeNode node, bool isLeft, int count) {
        if (node == null) {
            return count;
        }

        if (isLeft) {
            return Math.Max(DFS(node.left, true, 0), DFS(node.right, false, count+1));
        }
        return Math.Max(DFS(node.left, true, count+1), DFS(node.right, false, 0));
    }
}

File: C++/Binary Tree - DFS/1372. Longest ZigZag Path in a Binary Tree.cpp
/* 1372. Longest ZigZag Path in a Binary Tree
Difficulty: Medium
Submission Link: https://leetcode.com/problems/longest-zigzag-path-in-a-binary-tree/submissions/1388942924/ */

class Solution {
public:
    void sol(TreeNode* root,int l,int r,int& ans){
        if(root==NULL)return;
        ans=max(ans,max(l,r));
        sol(root->left,0,l+1,ans);
        sol(root->right,r+1,0,ans);
    }
    int longestZigZag(TreeNode* root) {
        int l=0,r=0,ans=0;
        sol(root,l,r,ans);
        return ans;
    }
};